xtask/src/task/build.rs

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{BuildArgs, BuildPkg, BuildProfile};
+use anyhow::Result;
+use itertools::Itertools;
+use log::debug;
+use std::{collections::BTreeSet, fmt::Display};
+use xshell::{cmd, Shell};
+
+/// Run the build command.
+pub fn run(args: BuildArgs) -> Result<()> {
+	debug!("rebuild targeting: {}", list_with_commas(&args.pkg));
+
+	let pkgs = args
+		.pkg
+		.into_iter()
+		.flat_map(resolve_packages)
+		.unique()
+		.collect::<BTreeSet<_>>();
+
+	// Each pkg produces two arguments, plus max two for the profile.
+	let mut builder_args = Vec::with_capacity((pkgs.len() * 2) + 2);
+
+	builder_args.extend(profile_args(&args.profile));
+
+	for pkg in &pkgs {
+		builder_args.extend([String::from("-p"), pkg.to_string()]);
+	}
+
+	debug!("rebuilding packages: {}", list_with_commas(&pkgs));
+	debug!("using profile: {}", args.profile.to_string());
+
+	let sh = Shell::new()?;
+	cmd!(sh, "cargo build {builder_args...}").run()?;
+
+	Ok(())
+}
+
+/// List a bunch of strings together separated by commas.
+fn list_with_commas(pkgs: impl IntoIterator<Item = impl ToString>) -> String {
+	pkgs.into_iter().map(|elem| elem.to_string()).join(", ")
+}
+
+/// Get the profile flags to use depending on the selection.
+fn profile_args(profile: &BuildProfile) -> impl IntoIterator<Item = String> {
+	match profile {
+		BuildProfile::Release => vec![String::from("--release")],
+		BuildProfile::Debug => vec![],
+		_ => vec![String::from("--profile"), profile.to_string()],
+	}
+}
+
+/// Resolve a set of packages based on the user's selection.
+fn resolve_packages(what: BuildPkg) -> Vec<Pkg> {
+	let mut vec = vec![];
+	extend_with_packages(what, &mut vec);
+	vec
+}
+
+/// Extend a vec with packages.
+fn extend_with_packages(what: BuildPkg, vec: &mut Vec<Pkg>) {
+	match what {
+		BuildPkg::All => {
+			extend_with_packages(BuildPkg::Core, vec);
+			extend_with_packages(BuildPkg::Sdk, vec);
+			extend_with_packages(BuildPkg::Plugins, vec);
+		}
+		_ => vec.extend(pkgs_for(what)),
+	}
+}
+
+/// Get the packages for a given selection.
+fn pkgs_for(what: BuildPkg) -> Vec<Pkg> {
+	match what {
+		BuildPkg::All => {
+			unimplemented!("should never be called with 'BuildWhat::Workspace'")
```

Comment:
Looks like this error message may have been missed during a rename of the enum variant to `BuildPkg::All`.

Reply:
Good catch! Pushed a fix.